File: Sources/BsnsSSHCore/SSHDecoder.swift
import Foundation

/// Reader counterpart to `SSHEncoder` for the SSH wire format (RFC 4251 §5).
/// Used to parse public-key blobs, signatures, and SSH-agent protocol
/// messages. Operates on a copied byte buffer so it is index-stable even when
/// constructed from a `Data` slice.
public struct SSHDecoder {
    public enum DecodeError: Error, Sendable {
        case truncated
        case invalidLength
    }

    private let bytes: [UInt8]
    private var offset = 0

    public init(_ data: Data) { self.bytes = [UInt8](data) }
    public init(_ bytes: [UInt8]) { self.bytes = bytes }

    public var isAtEnd: Bool { offset >= bytes.count }
    public var remaining: Int { bytes.count - offset }

    public mutating func readByte() throws -> UInt8 {
        guard offset < bytes.count else { throw DecodeError.truncated }
        defer { offset += 1 }
        return bytes[offset]
    }

    public mutating func readUInt32() throws -> UInt32 {
        guard offset + 4 <= bytes.count else { throw DecodeError.truncated }
        let value = (UInt32(bytes[offset]) << 24)
            | (UInt32(bytes[offset + 1]) << 16)
            | (UInt32(bytes[offset + 2]) << 8)
            | UInt32(bytes[offset + 3])
        offset += 4
        return value
    }

    public mutating func readBytes(_ count: Int) throws -> Data {
        guard count >= 0 else { throw DecodeError.invalidLength }
        guard offset + count <= bytes.count else { throw DecodeError.truncated }
        defer { offset += count }
        return Data(bytes[offset ..< offset + count])
    }

    /// SSH `string`: a `uint32` length followed by that many bytes.
    public mutating func readString() throws -> Data {
        let length = Int(try readUInt32())
        return try readBytes(length)
    }

    /// SSH `string` decoded as UTF-8.
    public mutating func readStringUTF8() throws -> String {
        let data = try readString()
        guard let string = String(data: data, encoding: .utf8) else {
            throw DecodeError.invalidLength
        }
        return string
    }
}

File: android/core/src/main/kotlin/cc/bsns/ssh/core/SshDecoder.kt
package cc.bsns.ssh.core

/**
 * Reader counterpart to [SshEncoder] for the SSH wire format (RFC 4251 §5).
 * Kotlin port of the Swift `SSHDecoder`; parses public-key blobs, signatures,
 * and SSH-agent messages.
 */
class SshDecoder(private val bytes: ByteArray) {
    class DecodeException(message: String) : Exception(message)

    private var offset = 0

    val isAtEnd: Boolean get() = offset >= bytes.size
    val remaining: Int get() = bytes.size - offset

    fun readByte(): Byte {
        if (offset >= bytes.size) throw DecodeException("truncated")
        return bytes[offset++]
    }

    /** Reads a 32-bit big-endian value into a Long (unsigned). */
    fun readUInt32(): Long {
        if (offset + 4 > bytes.size) throw DecodeException("truncated")
        val v = ((bytes[offset].toLong() and 0xFF) shl 24) or
            ((bytes[offset + 1].toLong() and 0xFF) shl 16) or
            ((bytes[offset + 2].toLong() and 0xFF) shl 8) or
            (bytes[offset + 3].toLong() and 0xFF)
        offset += 4
        return v
    }

    fun readBytes(count: Int): ByteArray {
        if (count < 0) throw DecodeException("invalidLength")
        if (offset + count > bytes.size) throw DecodeException("truncated")
        val out = bytes.copyOfRange(offset, offset + count)
        offset += count
        return out
    }

    /** SSH `string`: a uint32 length followed by that many bytes. */
    fun readString(): ByteArray = readBytes(readUInt32().toInt())

    fun readStringUtf8(): String = String(readString(), Charsets.UTF_8)
}
